[FEAT] 관리자 멘토링 목록 조회 페이지네이션

dye0p opened the pull request:
## Issue Number
closed #879 

## As-Is
<!-- 문제 상황 정의 -->
기존 관리자 페이지의 멘토링 목록 조회 기능에는 페이지네이션이 구현되어 있지 않아, 멘토링 데이터가 많아질 경우 성능 저하 및 사용성 문제가 발생할 수 있었습니다.

## To-Be
<!-- 변경 사항 -->
 ### 관리자 페이지의 멘토링 목록 조회에 페이지네이션 기능을 도입하였습니다.

- 멘토링 응답은 (식별자, 멘토 이름, 카테고리 이름리스트, 가격)입니다. (이 부분은 우선 기존에 보여지던 응답과 동일하게 구성하였습니다)

- 정렬 기준은 기본적으로 멘토링 id 내림차순으로 하였습니다.(최신 생성된것이 가장 먼저 보입니다)
- GET /admin/mentorings API에 page 파라미터를 추가하여 페이지별 조회가 가능하도록 수정했습니다.
- AdminMentoringService를 추가하여 관리자용 MentoringService를 구현할 수 있게 되었습니다.
- 페이지네이션 결과(PageResult)와 멘토링 정보(AdminMentoringResponse)를 위한 DTO를 추가했습니다.
- AdminMentoringServiceTest를 통해 페이지네이션 로직을 검증을 테스트 코드를 통해서 합니다.

### QueryDSL을 사용하여 MentoringRepository에 페이지네이션 조회 기능을 구현했습니다.
- Spring Data JPA의 기본 페이징 메서드를 사용하려 했지만, 카테고리 이름과 멘토이름을 가져오기 위해서 추가적인 조인 쿼리가 필요했기에 QueryDSL을 이용하여 구현했습니다.

## Check List
- [x] 테스트가 전부 통과되었나요?
- [x] 모든 commit이 push 되었나요?
- [x] merge할 branch를 확인했나요?
- [x] Assignee를 지정했나요?
- [x] Label을 지정했나요?
- [x] 닫을 이슈 번호를 지정했나요?

## (Optional) Additional Description
```java
public record AdminMentoringResponse(
        Long id,
        String mentorName,
        List<String> categories,
        int price
)
```
응답에 카테고리 이름이 리스트로 들어가야 했기에 QueryDSL의 `transform(gruopBy())` 메서드를 이용하려 했습니다. 
QueryDSL에서 사용하는 `Hibernate5Templates`과 스프링 부트에서 사용하는 `Hibernate6.x` 의 버전 불일치 문제로 `NoSuchMethodError` 예외가 발생하였습니다. 

몇가지 해결방법중 `JPAQueryFactory`를 생성할 때 QueryDSL이 `Hibernate5` 템플릿를 사용하지 않고, JPA의 JPQL 표준 템플릿을 사용하도록 명시해주었습니다. JPQL 표준은 Hibernate 템플릿의 내부구현에 의존하지 않기 때문에 문제가 해결 됩니다.

이 방법은 기존 Hibernate5Templates에서 사용하던 문법들을 사용하지 않고, JPQL 표준을 사용하기 때문에 `Hibernate5Templates`에서 사용하던 문법 사용시 문제가 될 수 있습니다.

그럼에도 이렇게 해결한 이유는 가장 간단하고 빠르게 해결할 수 있었기 때문이었습니다. 

이 이슈와 관련된 querydsl github 이슈 (해당 이슈에서 해결 방법을 찾을 수 있습니다)
- https://github.com/querydsl/querydsl/issues/3428

질문 있으시다면 코멘트에 남겨주세요!
  

<!-- This is an auto-generated comment: release notes by coderabbit.ai -->

## Summary by CodeRabbit

- 신규 기능
  - 관리자용 멘토링 목록 조회 기능 추가: 페이지네이션(페이지당 10건), 총 개수·총 페이지 제공, 최신순 정렬. 항목에는 멘토 이름·카테고리·가격 포함.
- 변경
  - 관리자 멘토링 삭제 관련 엔드포인트 경로를 단순화하여 일관성 개선.
- 테스트
  - 페이지 0~2 구간에서의 페이지네이션 동작(건수, 현재 페이지, 총 개수, 총 페이지) 검증 테스트 추가.

<!-- end of auto-generated comment: release notes by coderabbit.ai -->

dye0p committed:
remove: 사용하지 않는 클래스 제거

dye0p committed:
feat: 관리자가 필요한 멘토링 응답 DTO 추가

dye0p committed:
feat: 페이징 결과를 응답할 dto 추가

dye0p committed:
feat: 멘토링 페이징 및 전체 조회 기능 추가

- findAllWithPagination 메서드 추가하여 관리자용 전체 멘토링 조회 기능 구현
- AdminMentoringResponse 반환을 위한 JPQL 쿼리 작성 및 transform 사용

dye0p committed:
feat: QueryDslConfig JPAQueryFactory 생성시 JPQLTemplates 명시적 지정 추가

- Hibernate 버전 불일치 문제로 삭제된 메서드를 사용하려 했을때 예외가 발생하였음.
- QueryDsl JPAQueryFactory 클래스를 생성할 때 JPA 표준 문법을 사용할 수 있도록 명시적으로 Templates를 지정하였음

dye0p committed:
feat: 관리자 멘토링 Service 클래스 추가

- 관리자 별도로 멘토링에 관련된 service 로직을 수행할 수 있도록 관리자용 멘토링 service 구현

dye0p committed:
feat: 관리자 멘토링 조회 API 추가

dye0p committed:
test: FixtureUtil 코드 개선

- 테스트용 멘티와 멘토를 생성하는 메서드에서 전화번호를 9개까지만 유니크로 만들어서 10개를 생성하면 중복되는 전화번호가 생성되어 많은 픽스처를 만들지 못했음
- 전화번호 뒷자리를 0000~9999까지 만들 수 있도록 수정

dye0p committed:
test: 관리자 멘토링 전체 조회 페이지네이션 테스트 추가

dye0p commented:
@CodeRabbit 리뷰해

ppparkta commented on backend/fittoring/src/main/java/fittoring/admin/presentation/AdminMentoringController.java:
```diff
@@ -1,28 +1,50 @@
+    @GetMapping
+    public ResponseEntity<PageResult<AdminMentoringResponse>> getMentorings(
+            @Login LoginInfo loginInfo,
+            @RequestParam(defaultValue = "0") int page
```
p2: defaultValue 값이 0인데 다른분들이 작성해주신 기본값은 모두 1로 시작해서 통일해주시면 좋을 것 같습니다. 클라이언트에서 요청을 보낼 때도 1을 기준으로 보낼 것 같아요!

dye0p replied:
넵 수정하겠습니다!

ppparkta commented on backend/fittoring/src/main/java/fittoring/admin/service/AdminMentoringService.java:
```diff
@@ -0,0 +1,35 @@
+@Service
+public class AdminMentoringService {
+
+    private static final int PAGE_SIZE = 10;
```
p2: 이것도 통일을 위해서 20개로 수정해주시면 좋을 것 같습니다~

dye0p replied:
넵 수정하겠습니다!

jye326 commented on backend/fittoring/src/main/java/fittoring/admin/presentation/dto/AdminMentoringResponse.java:
```diff
@@ -0,0 +1,11 @@
+import java.util.List;
+
+public record AdminMentoringResponse(
+        Long id,
```
p3 : mentoring의 Id인가요? 그렇다면 클래스가 Mentoring이 아니니 `mentoringId`로 수정하는 게 어떨까요?

dye0p replied:
가독성에 더 좋을 것 같네요! 수정하겠습니다!

jye326 commented on backend/fittoring/src/main/java/fittoring/application/mentoring/repository/CustomMentoringRepositoryImpl.java:
```diff
@@ -1,46 +1,60 @@
-        SortKey sortKey,
-        Cursor cursor,
-        List<Long> categoryIds
+            SortKey sortKey,
+            Cursor cursor,
+            List<Long> categoryIds
     ) {
         // 결과 행을 튜플 타입(Mentoring, MentoringStatistics)으로 가져온다.
         List<Tuple> mentoringAndMentoringStatistics = jpaQueryFactory
                 .select(mentoring, mentoringStatistics)
                 .from(mentoring)
-                    .leftJoin(mentoringStatistics)
-                    .on(mentoringStatistics.mentoringId.eq(mentoring.id))
+                .leftJoin(mentoringStatistics)
+                .on(mentoringStatistics.mentoringId.eq(mentoring.id))
```
p3 : from절에 걸리는 join 부분은 들여쓰기가 있는 편이 가독성이 좋을 것 같은데 어떻게 생각하시나요?

dye0p replied:
아 자동정렬 때문에 수정되었네요..

jye326 commented on backend/fittoring/src/main/java/fittoring/application/mentoring/repository/CustomMentoringRepositoryImpl.java:
```diff
@@ -1,46 +1,60 @@
                 .from(mentoring)
-                    .leftJoin(mentoringStatistics)
-                    .on(mentoringStatistics.mentoringId.eq(mentoring.id))
+                .leftJoin(mentoringStatistics)
+                .on(mentoringStatistics.mentoringId.eq(mentoring.id))
                 .where(mentoringPaginationHelper.buildWhereClause(sortKey, cursor, categoryIds))
                 .orderBy(mentoringPaginationHelper.buildOrderSpecifiers(sortKey))
                 .limit(PAGE_SIZE + 1)
                 .fetch();
 
         // 다음 행이 존재할 경우 다음 커서를 문자열화 해서 반환한다.
         boolean hasNext = mentoringAndMentoringStatistics.size() > PAGE_SIZE;
-        String nextCursorCode = mentoringPaginationHelper.generateNextCursorCode(sortKey, hasNext, mentoringAndMentoringStatistics);
+        String nextCursorCode = mentoringPaginationHelper.generateNextCursorCode(
+                sortKey,
+                hasNext,
+                mentoringAndMentoringStatistics
+        );
```
👍

ppparkta commented on backend/fittoring/src/main/java/fittoring/application/mentoring/repository/CustomMentoringRepositoryImpl.java:
```diff
@@ -1,46 +1,60 @@
-import fittoring.application.mentoring.service.dto.MentoringPaginationResult;
+import fittoring.util.Cursor;
 import java.util.List;
 import lombok.RequiredArgsConstructor;
+import org.springframework.data.domain.Page;
+import org.springframework.data.domain.PageImpl;
+import org.springframework.data.domain.Pageable;
 
 @RequiredArgsConstructor
 public class CustomMentoringRepositoryImpl implements CustomMentoringRepository {
 
     private static final int PAGE_SIZE = 10;
     private static final QMentoring mentoring = QMentoring.mentoring;
     private static final QMentoringStatistics mentoringStatistics = QMentoringStatistics.mentoringStatistics;
+    private static final QCategoryMentoring categoryMentoring = QCategoryMentoring.categoryMentoring;
```
p2: 상수 컨벤션에 맞춰 대문자로 수정해주세요!

dye0p replied:
이 부분은 [[FEAT] 채팅기록 조회 API 구현](https://github.com/woowacourse-teams/2025-Fit-toring/pull/889#top)](https://github.com/woowacourse-teams/2025-Fit-toring/pull/889#discussion_r2442042945)
해당 코멘트에서 결정되었으므로 카멜케이스로 가겠습니다!

ppparkta commented on backend/fittoring/src/main/java/fittoring/application/mentoring/repository/CustomMentoringRepositoryImpl.java:
p3: 아까 이야기 나눈대로 admin 패키지 수정을 위해서 application 코드가 수정되는 것이 어색하게 느껴졌는데 패키지 분리가 완벽히 나눠지지 않아서 생기는 문제인 것 같아요. repository 인터페이스는 domain 패키지에, 구현체는 infrastructure에 두는 식으로 의존방향을 제어하는 리팩토링을 해보면 좋을 것 같아요!

아래의 글에서 이런 이야기를 다루고 있는데 흥미로운 얘기라 추천드려요! 멀티모듈을 도입해보자는 이야기는 아니고 패키지 관리를 어떻게 하는지 참고해보면 좋을 것 같아요!

> 저장소와 밀접한 중심 도메인을 다루는 계층은 더 견고하고 특별하게 격리되고 관리되어야 하기 때문에 반드시 분리되어야 합니다.

https://techblog.woowahan.com/2637/

dye0p replied:
저도 패키지 구조 문제에 동의합니다! 이 부분은 따로 이슈를 만들어서 구조 변경을 하면 좋을 것 같아요~ 좋은 의견 제안해주셔서 감사합니다 : )

jye326 commented on backend/fittoring/src/main/java/fittoring/application/mentoring/repository/CustomMentoringRepositoryImpl.java:
```diff
@@ -50,7 +64,54 @@ public MentoringPaginationResult findMentoringsWithPagination(
+                .offset(pageable.getOffset())
+                .limit(pageable.getPageSize())
+                .orderBy(mentoring.createdAt.desc())
+                .fetch();
```
p3 : 하나의 메서드에서 id 선 조회 및 최종 결과 조회의 두 작업을 차례로 수행하고 있네요. 메서드로 분리하는 것이 좋을 것 같아요!

dye0p replied:
그렇네요! 분리하겠습니다!

jye326 commented on backend/fittoring/src/main/java/fittoring/application/mentoring/repository/CustomMentoringRepositoryImpl.java:
```diff
@@ -50,7 +64,54 @@ public MentoringPaginationResult findMentoringsWithPagination(
+                .join(categoryMentoring.category)
+                .where(mentoring.id.in(ids))
+                .orderBy(mentoring.id.desc())
+                .transform(
+                        groupBy(mentoring.id)
+                                .list(
+                                        Projections.constructor(
+                                                AdminMentoringResponse.class,
+                                                mentoring.id,
+                                                mentoring.mentor.name,
+                                                GroupBy.list(categoryMentoring.category.title),
+                                                mentoring.price
+                                        )
+                                )
```
처음 봤는데 재밌는 기능이네요!

ppparkta reviewed (approved):
고생하셨습니다. 몇가지 코멘트 남겼는데 확인 부탁드립니다!

jye326 commented on backend/fittoring/src/main/java/fittoring/config/QueryDslConfig.java:
```diff
@@ -14,6 +15,6 @@ public class QueryDslConfig {
     @Bean
     public JPAQueryFactory jpaQueryFactory() {
-        return new JPAQueryFactory(entityManager);
+        return new JPAQueryFactory(JPQLTemplates.DEFAULT, entityManager);
```
p3: 어디서 버전 불일치 문제가 발생하고, Templates을 지정하는 것으로 어떻게 버전 불일치 문제가 해결되는 건지 궁금합니다!

> - Hibernate 버전 불일치 문제로 삭제된 메서드를 사용하려 했을때 예외가 발생하였음.
> - QueryDsl JPAQueryFactory 클래스를 생성할 때 JPA 표준 문법을 사용할 수 있도록 명시적으로 Templates를 지정하였음.

kimdevv replied:
p3: 저도 궁금합니다!

dye0p replied:
문서화가 필요할 것 같아서 문서화를 해두었습니다! 

[QueryDSL NoSuchMethod Excecption 발생 - 관리자 멘토링 페이지 네이션-](https://www.notion.so/QueryDSL-NoSuchMethod-Excecption-291cb79c55d080259e84d9805e2ca9ee?source=copy_link)

jye326 reviewed (approved):
수고하셨어요! 궁금한 점 몇 가지 코멘트 남기고 승인합니다~

kimdevv reviewed (commented):
고생하셨습니다~ 저도 간단한 커멘트 하나 남깁니다!

dye0p committed:
fix: 정렬기준 createdAt로 통일

dye0p committed:
refactor: 불필요한 조인을 제거하여 COUNT 쿼리 최적화

dye0p committed:
feat: id기반 내림차순 2차 정렬 조건 추가

- 같은 생성 시간의 경우 기존 페이지 개수를 초과한 결과가 반환되어서 제대로된 페이징이 되지 않는 문제가 있었음
- createAt의 값이 같은 경우 id기반으로 정렬을 수행할 수 있도록 2차 정렬조건을 orderBy에 추가 하였음

dye0p committed:
test: 정렬 검증 코드 추가

dye0p commented on backend/fittoring/src/main/java/fittoring/application/mentoring/repository/CustomMentoringRepositoryImpl.java:
```diff
@@ -50,7 +64,54 @@ public MentoringPaginationResult findMentoringsWithPagination(
+                                                mentoring.price
+                                        )
+                                )
+                );
```
createdAt을 기반으로 정렬 했을때 같은 시간인 경우 페이징 사이즈와 불일치 하게 결과가 반환되는 바람에, createdAt이 같을 경우 id 내림차순으로 정렬을 할 수 있도록 2차 정렬 기준을 orderBy()에 추가해줬어.

`.orderBy(mentoring.createdAt.desc(), mentoring.id.desc())`

dye0p commented on backend/fittoring/src/test/java/fittoring/admin/service/AdminMentoringServiceTest.java:
```diff
@@ -0,0 +1,116 @@
+                testAdmin.getId(), 0);
+        PageResult<AdminMentoringResponse> allForAdminPaged1 = adminMentoringService.findAllForAdminPaged(
+                testAdmin.getId(), 1);
+        PageResult<AdminMentoringResponse> allForAdminPaged2 = adminMentoringService.findAllForAdminPaged(
+                testAdmin.getId(), 2);
+
+        //then
+        SoftAssertions.assertSoftly(softly -> {
+            softly.assertThat(allForAdminPaged.content()).hasSize(10);
+            softly.assertThat(allForAdminPaged1.content()).hasSize(10);
+            softly.assertThat(allForAdminPaged2.content()).hasSize(10);
+
+            softly.assertThat(allForAdminPaged.page()).isEqualTo(0);
+            softly.assertThat(allForAdminPaged1.page()).isEqualTo(1);
+            softly.assertThat(allForAdminPaged2.page()).isEqualTo(2);
+
+            softly.assertThat(allForAdminPaged.total()).isEqualTo(30);
+            softly.assertThat(allForAdminPaged1.total()).isEqualTo(30);
+            softly.assertThat(allForAdminPaged2.total()).isEqualTo(30);
+
+            softly.assertThat(allForAdminPaged.totalPages()).isEqualTo(3);
+            softly.assertThat(allForAdminPaged1.totalPages()).isEqualTo(3);
+            softly.assertThat(allForAdminPaged2.totalPages()).isEqualTo(3);
+        });
```
좋다 추가했어!

dye0p commented on backend/fittoring/src/test/java/fittoring/admin/service/AdminMentoringServiceTest.java:
```diff
@@ -0,0 +1,116 @@
+        PageResult<AdminMentoringResponse> allForAdminPaged1 = adminMentoringService.findAllForAdminPaged(
+                testAdmin.getId(), 1);
+        PageResult<AdminMentoringResponse> allForAdminPaged2 = adminMentoringService.findAllForAdminPaged(
+                testAdmin.getId(), 2);
```
이것도 놓쳤네 고마워!

dye0p committed:
fix: 페이지 요청 기본값 0개에서 1개로 변경

dye0p committed:
refactor: 응답 멘토링 필드명 명확하게 변경

dye0p committed:
fix: 한 페이지당 응답하는 멘토링 개수 최대 20개로 수정

dye0p committed:
refactor: 멘토링 id 반환 메서드 분리

dye0p committed:
Merge branch 'develop' into feature/879-admin-mentorings-pagination

# Conflicts:
#	backend/fittoring/src/main/java/fittoring/admin/presentation/dto/PageResult.java
#	backend/fittoring/src/test/java/fittoring/application/FixtureUtil.java
